core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -161,28 +161,33 @@ public static void inputLongDecimal(Type type, LongDecimalWithOverflowAndLongSta
 
     private static void accumulateValueInState(Slice unscaledValue, LongDecimalWithOverflowAndLongState state)
     {
-        accumulateAndUpdateOverflow(unscaledValue, state);
-        state.setLong(state.getLong() + 1);
-    }
+        state.addLong(1); // row counter
 
-    private static void initializeIfNeeded(LongDecimalWithOverflowAndLongState state)
-    {
-        if (state.getLongDecimal() == null) {
-            state.setLongDecimal(UnscaledDecimal128Arithmetic.unscaledDecimal());
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
+            sum = UnscaledDecimal128Arithmetic.unscaledDecimal();
+            state.setLongDecimal(sum);
         }
+
+        long overflow = UnscaledDecimal128Arithmetic.addWithOverflow(sum, unscaledValue, sum);
+        state.addOverflow(overflow);
     }
 
     public static void combine(LongDecimalWithOverflowAndLongState state, LongDecimalWithOverflowAndLongState otherState)
     {
-        state.setLong(state.getLong() + otherState.getLong());
-        state.setOverflow(state.getOverflow() + otherState.getOverflow());
+        state.addLong(otherState.getLong()); // row counter
+
+        long overflow = otherState.getOverflow();
 
-        if (state.getLongDecimal() == null) {
+        Slice sum = state.getLongDecimal();
+        if (sum == null) {
             state.setLongDecimal(otherState.getLongDecimal());
```

Comment:
now updates in place will also affect `otherState.getLongDecimal()`. Should we make a copy here or is `otherState` discarded after `combine`?

Reply:
Same as above... I didn't change the original logic. I just restructured the code to avoid repeated calls to get state.

I'll leave that for a future revamp of this class.

Reply:
It’s risky and this is one of those cases where using slices that might be mutated it might be pointers into the middle of a very large allocation and therefore retain it (eg: VariableWidthBlock, the serialized block type of this state) are scary.

My understanding of this and the rest of the code is that this is safe because otherState is dead after combined with this state, and since the addInput call creates a new slice and doesn’t borrow from the input block when the current state is null, the second concern is also not a problem here. That said, this is indeed a fragile interaction. Not any more fragile than it was before, but hopefully something that can be improved with the future work.